src/update_roster.R

```diff
@@ -1,97 +1,276 @@
+            dplyr::arrange(gsis_id, status) |>
+            dplyr::distinct(gsis_id, .keep_all = TRUE),
           by = c("gsis_id"),
           na_matches = "never"
         ) |>
         fill_ids()
     }
-  }
 
-  if (season >= 2016 && nrow(weekly_rosters) == 0) {
-    roster <- shield
-  }
+    if (season >= 2016) {
+      weekly_rosters <- build_rosters_weekly_ngsapi(season)
+      if (nrow(weekly_rosters) > 0) {
+        roster <- convert_weekly_to_season_rosters(weekly_rosters)
+        shield <- build_rosters_shieldapi(season)
+        weekly_rosters <- weekly_rosters |>
+          dplyr::select(-c(height)) |>
+          dplyr::left_join(
+            shield |>
+              dplyr::select(gsis_id, birth_date, height) |>
+              dplyr::distinct(),
+            by = c("gsis_id"),
+            na_matches = "never"
+          ) |>
+          fill_ids()
+
+        roster <- roster |>
+          dplyr::select(-c(height)) |>
+          dplyr::left_join(
+            shield |>
+              dplyr::select(gsis_id, birth_date, height) |>
+              dplyr::distinct(),
+            by = c("gsis_id"),
+            na_matches = "never"
+          ) |>
+          fill_ids()
+      }
+    }
 
-  if ((season < 2002 && length(roster) == 0) | (season >= 2002 && length(weekly_rosters) == 0)) {
-    return(invisible(NULL))
-  }
+    if (season >= 2016 && nrow(weekly_rosters) == 0) {
+      roster <- shield
+    }
+
+    if ((season < 2002 &&
+         length(roster) == 0) |
+        (season >= 2002 && length(weekly_rosters) == 0)) {
+      return(invisible(NULL))
+    }
+
+    if (season >= 2002) {
+      weekly_rosters <- weekly_rosters |>
+        dplyr::mutate(
+          dplyr::across(
+            c(gsis_it_id,
+              smart_id),
+            as.character
+          ),
+          dplyr::across(
+            c(entry_year,
+              rookie_year),
+            as.integer
+          )
+        ) |>
+        dplyr::left_join(
+          nflreadr::load_players() |>
+            dplyr::select(
+              gsis_id,
+              full_name = display_name,
+              first_name,
+              last_name,
+              esb_id,
+              birth_date,
+              college = college_name,
+              entry_year,
+              rookie_year,
+              draft_club,
+              gsis_it_id,
+              smart_id,
+              height,
+              weight
+            ) |>
+            dplyr::mutate(
+              birth_date = as.Date(birth_date),
+              dplyr::across(
+                c(gsis_it_id,
+                  smart_id),
+                as.character
+              )
+            ),
+          by = c("gsis_id")
+        ) |>
+        dplyr::mutate(
+          dplyr::across(
+            dplyr::ends_with(".x"),
+            ~ dplyr::coalesce(get(
+              gsub("\\.x", ".y", dplyr::cur_column())
+            ), .),
+            .names = "{gsub('.x','',.col)}"
+          )
+        ) |>
```

Comment:
I wrote `join_coalesce()` https://nflreadr.nflverse.com/reference/join_coalesce.html a little while ago with this pattern in mind. Wondering if it would be a good fit here?

Reply:
Unfortunately, I don't think `join_coalesce` handles dates very well:

```r
df1 <- data.frame(
  name = c("Tom",
           "Bill",
           "Jen"),
  birth_date = c("1998-05-01",
                 "1997-12-12",
                 NA_character_)
) |>
  dplyr::mutate(birth_date = lubridate::as_date(birth_date))

df2 <- data.frame(
  name = c("Jen",
           "Tom",
           "Bill"),
  birth_date = c("1995-01-31",
                 NA_character_,
                 "1997-12-12")
) |>
  dplyr::mutate(birth_date = lubridate::as_date(birth_date))

nflreadr:::join_coalesce(
  df1,
  df2,
  by=c("name")
)
#>    name birth_date
#> 1: Bill      10207
#> 2:  Jen       9161
#> 3:  Tom      10347
```

Would it be alright if I kept my current coalesce logic in here and we can swap in `join_coalesce` when that's resolved?

Reply:
Yeah for sure. Can you drop this in nflreadr so I can look?